File: partiql-ast/src/main/java/org/partiql/ast/v1/Statement.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;

/**
 * TODO docs, equals, hashcode
 */
public abstract class Statement extends AstNode {
    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        if (this instanceof Query) {
            return visitor.visitQuery((Query) this, ctx);
        } else if (this instanceof DDL) {
            return visitor.visitDDL((DDL) this, ctx);
        } else if (this instanceof Explain) {
            return visitor.visitExplain((Explain) this, ctx);
        } else {
            throw new IllegalStateException("Unexpected value: " + this);
        }
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/Statement.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public abstract class Statement : AstNode() {
    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = when (this) {
        is Query -> visitor.visitQuery(this, ctx)
        is DDL -> visitor.visitDDL(this, ctx)
        is Explain -> visitor.visitExplain(this, ctx)
        else -> throw NotImplementedError()
    }
}
